Sources/Hyperspace/Service/Network/Cancellation.swift

```diff
@@ -0,0 +1,78 @@
+//
+//  Cancellation.swift
+//  Hyperspace
+//
+//  Created by Will McGinty on 2/19/19.
+//  Copyright © 2019 Bottle Rocket Studios. All rights reserved.
+//
+
+import Foundation
+
+/// Manages cancellation tokens and signals them when cancellation is requested.
+///
+/// All `CancellationTokenSource` methods are thread safe.
+public final class CancellationSource {
+    
+    public struct Token {
+        // MARK: Properties
+        private weak var source: CancellationSource?
+        public var isCancelling: Bool { return source?.isCancelling ?? false }
+        
+        public init(source: CancellationSource?) {
+            self.source = source
+        }
+        
+        // MARK: Interface
+        public func register(closure: @escaping () -> Void) {
+            source?.register(closure)
+        }
+    }
+    
+    // MARK: Properties
+    private let lock = NSLock()
+    private var observers: [() -> Void]? = []
```

Comment:
What's the reasoning around this being an optional property that gets immediately set to an empty array?

Reply:
They are used to represent different states - the initial state is an empty array to indicate their own observers. Upon cancellation we nil the array out and then use that `.none` state to indicate that cancellation of this source has already occurred, so any blocks that are registered from that point on will run immediately.

It's not super clear I agree, but a) this shouldn't really need to be 'understood' by the user b) it's going to improve performance a bit locking the single array vs locking the array and an `isCancelling` bool to keep state.